File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-3.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @param {number} k
     * @return {ListNode}
     */
    reverseKGroup(head, k) {
        const dummy = new ListNode(0, head);
        let prevGroup = dummy;
        while (true) {
            let kNode = prevGroup;
            for (let i = 0; i < k; ++i) {
                kNode = kNode.next;
                if (!kNode) return dummy.next;
            }
            let nextGroup = kNode.next;
            let prev = kNode.next;
            let curr = prevGroup.next;
            while (curr != nextGroup) {
                let next = curr.next;
                curr.next = prev;
                prev = curr;
                curr = next;
            }
            let tmp = prevGroup.next;
            prevGroup.next = prev;
            prevGroup = tmp;
        }
        return dummy.next;
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        # Get center node
        slow = head
        fast = head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        # Reverse right-half list
        node = slow.next
        prev = slow.next = None
        while node:
            node_next = node.next
            node.next = prev
            prev = node
            node = node_next
        # Cross linked for reordering list
        first = head
        second = prev
        while second:
            node = second.next
            second.next = first.next
            first.next = second
            first = second.next
            second = node